tracee-rules/input.go

```diff
@@ -4,70 +4,174 @@ import (
 	"bufio"
 	"encoding/gob"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io"
 	"log"
 	"os"
-	"time"
+	"strings"
 
 	"github.com/aquasecurity/tracee/tracee-rules/types"
 	tracee "github.com/aquasecurity/tracee/tracee/external"
 )
 
-func setupStdinSource(inputSource string) (chan types.Event, error) {
-	res := make(chan types.Event)
-	scanner := bufio.NewScanner(os.Stdin)
-	go func() {
-		for scanner.Scan() {
-			event := scanner.Bytes()
-			switch inputSource {
-			case "tracee":
-				var e tracee.Event
-				err := json.Unmarshal(event, &e)
-				if err != nil {
-					log.Printf("invalid json in %s: %v", string(event), err)
-				}
-				res <- types.Event(e)
-			}
-		}
-		close(res)
-	}()
-	return res, nil
+var errHelp = errors.New("user has requested help text")
+
+type inputFormat uint8
+
+const (
+	invalidInputFormat inputFormat = iota
+	jsonInputFormat
+	gobInputFormat
+)
+
+type traceeInputOptions struct {
+	inputFile   *os.File
+	inputFormat inputFormat
 }
 
-func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
-	_, err := os.Stat(traceeFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid Tracee input file: %s", traceeFilePath)
+func setupTraceeInputSource(opts *traceeInputOptions) (chan types.Event, error) {
+
+	if opts.inputFormat == jsonInputFormat {
+		return setupTraceeJSONInputSource(opts)
 	}
-	f, err := os.Open(traceeFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid file: %s", traceeFilePath)
+
+	if opts.inputFormat == gobInputFormat {
+		return setupTraceeGobInputSource(opts)
 	}
-	dec := gob.NewDecoder(f)
+
+	return nil, errors.New("could not set up input source")
+}
+
+func setupTraceeGobInputSource(opts *traceeInputOptions) (chan types.Event, error) {
+	dec := gob.NewDecoder(opts.inputFile)
 	res := make(chan types.Event)
 	go func() {
 		for {
 			var event tracee.Event
 			err := dec.Decode(&event)
 			if err != nil {
 				if err == io.EOF {
-					// ignore EOF because we assume events can keep streaming into the file
-					// this might create a backlog of events and depending on the implementation of the input source might lead to lost events
-					// TODO: investigate impact of this and research alternatives
-					time.Sleep(time.Millisecond * 150)
-					continue
+					break
 				} else {
 					log.Printf("Error while decoding event: %v", err)
```

Comment:
Is it possible for the input file to be closed when we get here? If that happens we will be stuck in a loop forever.

Reply:
(I realize it's not a change of logic here but it's just me looking at this with a fresh set of eyes)

Reply:
Hm, that is a good point but i'm not sure how it would?